source/extensions/http/header_validators/envoy_default/header_validator.cc

```diff
@@ -8,38 +14,344 @@ namespace EnvoyDefault {
+      "UPDATEREDIRECTREF",
+      "VERSION-CONTROL",
+      "*",
+  };
+
+  const auto& method = value.getStringView();
+  bool is_valid = true;
 
-HeaderValidatorFactory::HeaderValidatorFactory(const HeaderValidatorConfig& config)
-    : config_(config) {}
+  if (config_.restrict_http_methods()) {
+    is_valid = kHttpMethodRegistry.contains(method);
+  } else {
+    is_valid = !method.empty();
+    for (std::size_t i = 0; i < method.size() && is_valid; ++i) {
+      is_valid = testChar(kMethodHeaderCharTable, method.at(i));
+    }
+  }
 
-::Envoy::Http::HeaderValidatorPtr
-HeaderValidatorFactory::create(::Envoy::Http::Protocol protocol,
-                               StreamInfo::StreamInfo& stream_info) {
-  return std::make_unique<HeaderValidator>(config_, protocol, stream_info);
+  if (!is_valid) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidMethod};
+  }
+
+  return HeaderEntryValidationResult::success();
 }
 
-HeaderValidator::HeaderValidator(const HeaderValidatorConfig&, ::Envoy::Http::Protocol,
-                                 StreamInfo::StreamInfo& stream_info)
-    : stream_info_(stream_info) {}
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+HeaderValidator::validateSchemeHeader(const HeaderString& value) {
+  //
+  // From RFC 3986, https://datatracker.ietf.org/doc/html/rfc3986#section-3.1:
+  //
+  // scheme = ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )
+  //
+  // Although schemes are case-insensitive, the canonical form is lowercase and documents that
+  // specify schemes must do so with lowercase letters. An implementation should accept uppercase
+  // letters as equivalent to lowercase in scheme names (e.g., allow "HTTP" as well as "http") for
+  // the sake of robustness but should only produce lowercase scheme names for consistency.
+  //
+  // The validation mode controls whether uppercase letters are permitted.
+  //
+  const auto& value_string_view = value.getStringView();
+
+  if (value_string_view.empty()) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidScheme};
+  }
+
+  auto character_it = value_string_view.begin();
+
+  // The first character must be an ALPHA
+  auto valid_first_character = (*character_it >= 'a' && *character_it <= 'z') ||
+                               (*character_it >= 'A' && *character_it <= 'Z');
+  if (!valid_first_character) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidScheme};
+  }
+
+  for (++character_it; character_it != value_string_view.end(); ++character_it) {
+    if (!testChar(kSchemeHeaderCharTable, *character_it)) {
+      return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidScheme};
+    }
+  }
+
+  return HeaderEntryValidationResult::success();
+}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+HeaderValidator::validateStatusHeader(const StatusPseudoHeaderValidationMode& mode,
+                                      const HeaderString& value) {
+  //
```

Comment:
nit: not sure why comment blocks start and end with empty comment line but I'd prefer we remove those